File: PROGRAMMERS/LEVEL_02/Q17680.cpp
//https://programmers.co.kr/learn/courses/30/lessons/17680
//캐시
#include <iostream>

#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(int cacheSize, vector<string> cities) {

    const int HIT = 1;
    const int MISS = 5;
    
    // 캐시 사이즈가 0 일경우
    if (cacheSize == 0)
    {
        return cities.size() * MISS;
    }

    // 캐시를 저장할 벡터, 시간 경과를 저장할 벡터 선언
    vector<string> q(cacheSize);

    int time = 0;
    for (auto iter_city = cities.begin(); iter_city != cities.end(); ++iter_city)
    {
        // 문자를 대문자로 변환
        transform(iter_city->begin(), iter_city->end(), iter_city->begin(), ::toupper);

        // cache벡터에서 해당 도시 찾기
        auto iter_cache = find_if(q.begin(), q.end(), [&iter_city](const string& s) ->bool {
            return s.compare(*iter_city) == 0;
            });
        
        // cache벡터에서 도시를 찾았을 때
        if (iter_cache != q.end())
        {
            q.erase(iter_cache);
            q.push_back(*iter_city);

            time += HIT;
        }
        // cache벡터에 도시가 없을 때
        else
        {
            if (q.size() >= cacheSize)
            {
                q.erase(q.begin());
            }
            if (q.size() < cacheSize)
            {
                q.push_back(*iter_city);
            }

            time += MISS;
        }
    }

    return time;
}

int main()
{
    /*
    int cacheSize = 3; vector<string> cities = { "Jeju", "Pangyo", "Seoul", "Jeju", "Pangyo", "Seoul", "Jeju", "Pangyo", "Seoul" };
    int cacheSize = 3; vector<string> cities = { "Jeju", "Pangyo", "Seoul", "NewYork", "LA", "Jeju", "Pangyo", "Seoul", "NewYork", "LA" };
    int cacheSize = 2; vector<string> cities = { "Jeju", "Pangyo", "Seoul", "NewYork", "LA", "SanFrancisco", "Seoul", "Rome", "Paris", "Jeju", "NewYork", "Rome" };
    int cacheSize = 0; vector<string> cities = { "Jeju", "Pangyo", "Seoul", "NewYork", "LA"};
    int cacheSize = 2; vector<string> cities = { "Jeju", "Pangyo", "NewYork", "NewYork"};
    */
    int cacheSize = 5; vector<string> cities = { "Jeju", "Pangyo", "Seoul", "NewYork", "LA", "SanFrancisco", "Seoul", "Rome", "Paris", "Jeju", "NewYork", "Rome" };

    cout << solution(cacheSize, cities);

    return 0;
}

File: PROGRAMMERS/LEVEL_02/Q17680.java
package PROGRAMMERS.LEVEL_02;

import java.util.LinkedList;
import java.util.Queue;

public class Q17680 {

    public static void main(String[] args) {
        /*
        도시이름(cities)	                                        캐시크기(cacheSize)	실행시간
        ["Jeju", "Pangyo", "Seoul", "NewYork", "LA",
            "Jeju", "Pangyo", "Seoul", "NewYork", "LA"]	        3                   50
        ["Jeju", "Pangyo", "Seoul", "Jeju", "Pangyo",
            "Seoul", "Jeju", "Pangyo", "Seoul"]	                3                   21
        ["Jeju", "Pangyo", "Seoul", "NewYork", "LA",
            "SanFrancisco", "Seoul", "Rome", "Paris", "Jeju",
            "NewYork", "Rome"]	                                2                   60
        ["Jeju", "Pangyo", "Seoul", "NewYork", "LA",
            "SanFrancisco", "Seoul", "Rome", "Paris", "Jeju",
            "NewYork", "Rome"]	                                5                   52
        ["Jeju", "Pangyo", "NewYork", "newyork"]	            2                   16
        ["Jeju", "Pangyo", "Seoul", "NewYork", "LA"]	        0                   25
         */

        String[] cities = {"Jeju", "Pangyo", "Seoul", "Jeju", "Pangyo", "Seoul", "Jeju", "Pangyo", "Seoul"};
        int cacheSize = 3;
        System.out.println(solution(cacheSize, cities));
    }

    public static int solution(int cacheSize, String[] cities) {

        final int HIT = 1;
        final int MISS = 5;
        if (cacheSize == 0) {
            return cities.length * MISS;
        }

        int time = 0;
        Queue<String> queue = new LinkedList<>();

        for (String city : cities) {
            city = city.toUpperCase();

            if (queue.remove(city)) {

                time += HIT;
            } else {

                time += MISS;
                if (queue.size() == cacheSize) {
                    queue.poll();
                }
            }
            queue.add(city);
        }

        return time;
    }
}
